app/src/main/java/woowacourse/movie/view/home/MovieAdapter.kt

```diff
@@ -0,0 +1,90 @@
+package woowacourse.movie.view.home
+
+import android.view.LayoutInflater
+import android.view.ViewGroup
+import androidx.databinding.DataBindingUtil
+import androidx.recyclerview.widget.DiffUtil
+import androidx.recyclerview.widget.ListAdapter
+import androidx.recyclerview.widget.RecyclerView
+import woowacourse.movie.R
+import woowacourse.movie.databinding.ItemMovieBinding
+import woowacourse.movie.model.movie.Movie
+
+class MovieAdapter(
+    private val movieClickListener: MovieClickListener,
+    private val advertisementClickListener: () -> Unit,
+) : ListAdapter<Movie, RecyclerView.ViewHolder>(
+        object : DiffUtil.ItemCallback<Movie>() {
+            override fun areItemsTheSame(
+                oldItem: Movie,
+                newItem: Movie,
+            ): Boolean = oldItem.id == newItem.id
+
+            override fun areContentsTheSame(
+                oldItem: Movie,
+                newItem: Movie,
+            ): Boolean = oldItem == newItem
+        },
+    ) {
+    override fun getItemViewType(position: Int): Int =
+        when {
+            (position + 1) % AD_POSITION_MULTIPLE == 0 -> AD_ITEM_TYPE
+            else -> MOVIE_ITEM_TYPE
+        }
+
+    override fun onCreateViewHolder(
+        parent: ViewGroup,
+        viewType: Int,
+    ): RecyclerView.ViewHolder =
+        when (viewType) {
+            MOVIE_ITEM_TYPE -> {
+                val inflater = LayoutInflater.from(parent.context)
+                val movieBinding =
+                    DataBindingUtil.inflate<ItemMovieBinding>(
+                        inflater,
+                        R.layout.item_movie,
+                        parent,
+                        false,
+                    )
+                val holder = MovieViewHolder(movieBinding)
+                holder.button.setOnClickListener {
+                    val position = holder.adapterPosition
+                    val adjustedPosition = position - position / AD_POSITION_MULTIPLE
+                    val item = getItem(adjustedPosition)
+                    movieClickListener.onReservationClick(item.id)
+                }
+                holder
+            }
```

Comment:
onCreateViewHolder에서 이렇게만 남기도록 구현해보시면 어떨까요?
```suggestion
            MOVIE_ITEM_TYPE -> { 
                MovieViewHolder(movieBinding)
            }
```

Reply:
팩토리 함수를 만들어서 clickListener 의 함수와 parent 만 넘겨 객체를 초기화 할 수 있도록 수정했습니다!
cf000c0530f9b72e61383a719c5456af3e5b7e3c (해당 영화 목록 리사이클러뷰 로직을 전체적으로 개선한 커밋입니다)